[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)

## Overview of an example workflow: Fastq data quality checking

This is an example workflow to check the quality of the paired-end fastq files using `FastQC` software.

![](graphs/diagram.png)

## Installation

- __Running environment__: 
    - The workflow was constructed based on macOS Catalina 10.15.7 running the Oracle v1.8 java runtime environment (JREs). However, you can also run this using your preferred Linux distribution.

- __Required software and versions__: 
    - [Trim_galore! v0.6.4_dev](https://github.com/FelixKrueger/TrimGalore) 
    - [Cutadapt v1.8.1](https://cutadapt.readthedocs.io/en/stable/)
    - [fastQC v0.11.5](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
    - [BSMAP v2.74](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-232)
    - [samtools v1.3](https://github.com/samtools/samtools)
    - [bamtools v2.4.0](https://github.com/pezmaster31/bamtools/)
    - [Java v1.8.0_45](https://www.oracle.com/java/technologies/downloads/) 
    - [Picard v2.9.0](https://broadinstitute.github.io/picard/)
    - [bamUtil v1.0.13](https://genome.sph.umich.edu/wiki/BamUtil)
    - [Python v2.7.5](https://www.python.org/)
    - [bedGraphToBigWig](http://hgdownload.soe.ucsc.edu/admin/exe/)
    - [perl v5.26.2](https://www.perl.org/)
    - [IGV v2.5.3](https://software.broadinstitute.org/software/igv/)
    - [R v4.1](https://www.r-project.org/)
        - [tidyverse v2.0.0](https://github.com/tidyverse/tidyverse)

## Input Data

The example data used here is the paired-end fastq file generated by using the Illumina platform.  

- R1 FASTQ file: `input/B73_chr1_subset_reads_1.fastq`  
- R2 FASTQ file: `input/B73_chr1_subset_reads_2.fastq`  

Each entry in a FASTQ files consists of 4 lines:  

1. A sequence identifier with information about the sequencing run and the cluster. The exact contents of this line vary by based on the BCL to FASTQ conversion software used.  
2. The sequence (the base calls; A, C, T, G and N).  
3. A separator, which is simply a plus (+) sign.  
4. The base call quality scores. These are Phred +33 encoded, using ASCII characters to represent the numerical quality scores.  

The first entry of the input data:
```
@SRR8738272.153232
TGATTTGAAATTAAACGAATATGGAAATCGGTTTGAAGGTTTTGGAATCGAGTATAATTGGATTTACAAATGTGGTTTATGGGAATTTTTTTATGTGAAAGTTTTGATTCTGATGTATAATATTGA
+
CCCCCGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG@
```


## Major steps

#### Step 1: Trimming the reads and running FastQC for quality checking
- Note that you have to normalize the path in the shell script.

```
sh workflow/1_trim_reads.sh
```

#### Step 2: Mapping reads using BSMAP

```
sh workflow/2_map_reads.sh
```

#### Step 3: View the results

- Results can be converted into a bigWig format, which can be visualized using IGV.

```
3_visualize_results.sh
```

#### Step 4: Identify unmethylated regions


```
4_find_UMRs.sh
```

## Expected results

![](graphs/figure1.png)

## License
It is a free and open source software, licensed under [GPLv3](https://github.com/github/choosealicense.com/blob/gh-pages/_licenses/gpl-3.0.txt).
